### PR TITLE
✨  (go/v3) change the Dockerfile to allow built the Operator manager for different platforms

### PR DESCRIPTION
**Pull request by gquillar**
Currently the binary in the go/v3 scaffold template images is built using amd64 go arch.
One of its motivations is to provide support for ppc64le and s390x arches.

Signed-off-by: Gilles Quillard <gquillar@us.ibm.com>

### CODE REVIEW

**Review by camilamacedo86 — COMMENTED** [review]
/ok-to-test

@gquillar :

a) you need to assign to the CLA
b) The title emoji is without (). Note that it is used to build the changelog, so we also test it in the ci

File: pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:133]: I do not think that we can change in this way.

Note that: 
I am Operator Author who is working in darwin/amd64 local env. However, my operator will be published on Linux/amd64 cluster which is the default/common scenario. So if we move with your change it would fail because the GOARCH used is not the default. 

I think we can: 
Create a variable in the Makefile with the GOARCH fixed for amd64 
then we can comment on top to let authors know if they want to build the Operator manager image for another arch then can update that.

File: pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go:44]: If we change the value informed by the ` --build-arg GOARCH=$(shell go env GOARCH) ` will it work or we will build to amd64 always? 
Should we not create a parameter to receive the OS as well?

File: pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go
Comment (gquillar) [reply pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:133]: Okay, I can create 2 variables for the target OS and ARCH and set them by default at linux and amd64:
```
TARGET_GOOS   ?= linux
TARGET_GOARCH ?= amd64
```
Then they can be passed as build arguments to the Dockerfile:
`docker build -t ${IMG} --build-arg TARGET_GOOS=${TARGET_GOOS} --build-arg TARGET_GOARCH=${TARGET_GOARCH} .`

**Review by camilamacedo86 — CHANGES_REQUESTED** [review]
This PR needs to sort out the lint issues and be rabased with the master.
I think we can move with this one but we will also need a follow looking to check if we have an option which addresses better the needs to support multi-platforms.

**Comment by aemperador** [comment]
hi @camilamacedo86, Gilles has been out on sick leave, so I will be finishing out this PR. I rebased, and have filled out an EasyCLA application. is there anything else that needs to be done?

**Comment by aemperador** [comment]
@camilamacedo86 I have rebased the PR and filled out the EasyCLA, but it is still showing that I haven't filled out EasyCLA. please advise

File: pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:68]:
```suggestion
# TARGET_GOOS and TARGET_GOARCH define respectively the OS and architecture used to build
# the manager image.
```

File: pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go:47]: /hold

based on : https://github.com/kubernetes-sigs/kubebuilder/issues/2850

We need to see what is the best way / approach to help those that want to build images for multi-archs

**Comment by jaypoulz** [comment]
Hi @camilamacedo86  - is this PR made redundant by your work in https://github.com/kubernetes-sigs/kubebuilder/pull/2906?

**Comment by camilamacedo86** [comment]
Hi [jaypoulz](https://github.com/jaypoulz), 

> Hi @camilamacedo86 - is this PR made redundant by your work in https://github.com/kubernetes-sigs/kubebuilder/pull/2906?

This PR does not address properly the problem. 
So, acctually is the opposite. The work done https://github.com/kubernetes-sigs/kubebuilder/pull/2906 (which was also motivated by this PR) when merged would also close this one.

File: pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/dockerfile.go:63]: Docker has ARGS defined for both already which are TARGETOS and TARGETARCH

File: pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go
Comment (camilamacedo86) [inline pkg/plugins/golang/v3/scaffolds/internal/templates/makefile.go:138]: By using the ARGs provided by docker we do not need to pass them here.

File: testdata/project-v3-addon-and-grafana/Dockerfile
Comment (camilamacedo86) [inline testdata/project-v3-addon-and-grafana/Dockerfile:7]: same we do not need to define it

**Review by camilamacedo86 — COMMENTED** [review]
The work done https://github.com/kubernetes-sigs/kubebuilder/pull/2906 (which was also motivated by this PR) when merged would also close this one.

/hold

**Comment by jaypoulz** [comment]
Perfect - yes - that's exactly what I was asking. :) We have an internal tracker for this PR, but I figured we might be able to close it in favor of #2906. Sounds like that is the case!

**Comment by aemperador** [comment]
hi @camilamacedo86, Gilles is still out on sick leave. I reviewed #2906, lgtm 👍 we can close this once that's merged and available

**Pull request closed without merging**
